src/Symfony/Component/Filesystem/Path.php

```diff
@@ -0,0 +1,990 @@
+     * path.
+     *
+     * This method is able to deal with both UNIX and Windows paths.
+     *
+     * @param string $path A path string.
+     *
+     * @return string The normalized path.
+     *
+     * @since 2.2 Added method.
+     */
+    public static function normalize(string $path): string
+    {
+        return str_replace('\\', '/', $path);
+    }
+
+    /**
+     * Returns the directory part of the path.
+     *
+     * This method is similar to PHP's dirname(), but handles various cases
+     * where dirname() returns a weird result:
+     *
+     *  - dirname() does not accept backslashes on UNIX
+     *  - dirname("C:/webmozart") returns "C:", not "C:/"
+     *  - dirname("C:/") returns ".", not "C:/"
+     *  - dirname("C:") returns ".", not "C:/"
+     *  - dirname("webmozart") returns ".", not ""
+     *  - dirname() does not canonicalize the result
+     *
+     * This method fixes these shortcomings and behaves like dirname()
+     * otherwise.
+     *
+     * The result is a canonical path.
+     *
+     * @param string $path A path string.
+     *
+     * @return string The canonical directory part. Returns the root directory
+     *                if the root directory is passed. Returns an empty string
+     *                if a relative path is passed that contains no slashes.
+     *                Returns an empty string if an empty string is passed.
+     *
+     * @since 1.0 Added method.
+     * @since 2.0 Method now fails if $path is not a string.
+     */
+    public static function getDirectory(string $path): string
+    {
+        if ('' === $path) {
+            return '';
+        }
+
+        $path = static::canonicalize($path);
+
+        // Maintain scheme
+        if (false !== ($pos = strpos($path, '://'))) {
+            $scheme = substr($path, 0, $pos + 3);
+            $path = substr($path, $pos + 3);
+        } else {
+            $scheme = '';
+        }
+
+        if (false !== ($pos = strrpos($path, '/'))) {
+            // Directory equals root directory "/"
+            if (0 === $pos) {
+                return $scheme.'/';
+            }
+
+            // Directory equals Windows root "C:/"
+            if (2 === $pos && ctype_alpha($path[0]) && ':' === $path[1]) {
+                return $scheme.substr($path, 0, 3);
+            }
+
+            return $scheme.substr($path, 0, $pos);
+        }
+
+        return '';
+    }
+
+    /**
+     * Returns canonical path of the user's home directory.
+     *
+     * Supported operating systems:
+     *
+     *  - UNIX
+     *  - Windows8 and upper
+     *
+     * If your operation system or environment isn't supported, an exception is thrown.
+     *
+     * The result is a canonical path.
+     *
+     * @return string The canonical home directory
+     *
+     * @throws RuntimeException If your operation system or environment isn't supported
+     *
+     * @since 2.1 Added method.
+     */
+    public static function getHomeDirectory(): string
```

Comment:
should this really be part of this class? It feels like something _some_ users will need, also all other methods are working on the provided path, while this returns home directory

Reply:
This should probably be on `Filesystem` class

Reply:
The method is required in order to correctly resolve `~` to the path of the home directory.

You could move this to `Filesystem`, but then it would need to remain static in order to be usable here. Static methods, however, go against the conventions of `Filesystem`, where the assumption is to create an instance first. Last, one could instantiate a `Filesystem` instance in here every time this is needed, but I rather advice against that. This class is a very very low-level building block and designed for speed, since there could be thousands of calls to the methods of this class.

Reply:
I understand, just feels weird as it doesn't fit with all the other methods on this class.